File: sugars/rust/src/safety.rs

// No type error - not type safe!
macro_rules! ill_typed {
    () => {
        1 + "two";
    }
}

// No unbound id error - not scope safe.
macro_rules! ill_scoped {
    () => ({
        x
    })
}

// Rust macros are hygienic
macro_rules! hygienic {
    ($i:ident, $body:expr) => {
        let x = "macro";
        let $i = "user";
        println!("macro->{}", x);
        let x = "macro";
        $body;
    }
}

// This macro has a lurking syntax error, which won't be found until it is used.
macro_rules! syntax_unsafe {
    ($e:expr) => {
        let $e = 0; // Error: e is an expr, not a pattern!
        ()
    }
}

pub fn safety() {
    hygienic!(x, println!("user->{}", x));
    // syntax_unsafe!(1 + 2); // Syntax error when macro is used
    // ill_typed!(); // Type error when macro is used
}

File: sugars/cpp/safety.c
#include <stdio.h>

// Not syntax safe
#define SUB1(a, b) a - b
#define SUB2(a, b) (a) - (b)
#define SUB3(a, b) ((a) - (b))

// No error: not scope safe.
#define ILL_SCOPED x

// No type error: not type safe.
#define ILL_TYPED int* ptr = 2.3

// No hygienic
#define UNHYGIENIC x = x + 1

int main() {
  //ILL_TYPED;
  int x = 0;
  UNHYGIENIC;
  printf("Unhygienic: %i\n", x);
  // Not syntax safe!
  printf("'#define SUB(a, b) a - b'       in SUB(0, 2-1): %i\n", SUB1(0, 2-1));
  printf("'#define SUB(a, b) (a) - (b)'   in SUB(5, 3)*2: %i\n", SUB2(5, 3)*2);
  printf("'#define SUB(a, b) ((a) - (b))' in SUB(0, 2-1): %i\n", SUB3(0, 2-1));
  printf("'#define SUB(a, b) ((a) - (b))' in SUB(5, 3)*2: %i\n", SUB3(5, 3)*2);
  return 0;
}
